File: src/main/java/org/tasks/reminders/NotificationDialog.java
package org.tasks.reminders;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;

import org.tasks.R;
import org.tasks.dialogs.DialogBuilder;
import org.tasks.injection.InjectingDialogFragment;

import java.util.List;

import javax.inject.Inject;

import static java.util.Arrays.asList;

public class NotificationDialog extends InjectingDialogFragment {

    public interface NotificationHandler {
        void edit();

        void snooze();

        void complete();

        void dismiss();
    }

    @Inject DialogBuilder dialogBuilder;

    private String title;
    private NotificationHandler handler;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        List<String> items = asList(
                getString(R.string.TAd_actionEditTask),
                getString(R.string.rmd_NoA_snooze),
                getString(R.string.rmd_NoA_done));

        handler = (NotificationHandler) getActivity();

        return dialogBuilder.newDialog()
                .setTitle(title)
                .setItems(items.toArray(new String[items.size()]), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        switch (which) {
                            case 0:
                                handler.edit();
                                break;
                            case 1:
                                handler.snooze();
                                break;
                            case 2:
                                handler.complete();
                                break;
                        }
                    }
                })
                .show();
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        handler.dismiss();
    }

    public void setTitle(String title) {
        this.title = title;
        Dialog dialog = getDialog();
        if (dialog != null) {
            dialog.setTitle(title);
        }
    }
}

File: app/src/main/java/org/tasks/reminders/NotificationDialog.kt
package org.tasks.reminders

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import dagger.hilt.android.AndroidEntryPoint
import org.tasks.R
import org.tasks.dialogs.DialogBuilder
import org.tasks.reminders.NotificationActivity.Companion.EXTRA_COMPLETE_LABEL
import org.tasks.reminders.NotificationActivity.Companion.EXTRA_SNOOZE_LABEL
import org.tasks.reminders.NotificationActivity.Companion.EXTRA_READ_ONLY
import javax.inject.Inject

@AndroidEntryPoint
class NotificationDialog : DialogFragment() {
    @Inject lateinit var dialogBuilder: DialogBuilder

    private var title: String? = null
    private lateinit var handler: NotificationHandler

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        handler = activity as NotificationHandler
        val readOnly = requireArguments().getBoolean(EXTRA_READ_ONLY)

        val completeLabel = requireArguments().getString(EXTRA_COMPLETE_LABEL)
        val snoozeLabel = requireArguments().getString(EXTRA_SNOOZE_LABEL)
        val actions = buildList<Pair<String, () -> Unit>> {
            add(getString(R.string.TAd_actionEditTask) to handler::edit)
            add(snoozeLabel.orEmpty() to handler::snooze)
            if (!readOnly) {
                add(completeLabel.orEmpty() to handler::complete)
            }
        }
        return dialogBuilder
            .newDialog(title)
            .setItems(actions.map { it.first }) { _, which -> actions[which].second() }
            .setNegativeButton(R.string.cancel) { _, _ -> handler.dismiss() }
            .show()
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)

        handler.dismiss()
    }

    fun setTitle(title: String?) {
        this.title = title
        dialog?.setTitle(title)
    }

    interface NotificationHandler {
        fun edit()

        fun snooze()

        fun complete()

        fun dismiss()
    }
}
